library/src/G4CMPPhononBoundaryProcess.cc

```diff
@@ -410,3 +419,50 @@ GetLambertianVector(const G4ThreeVector& surfNorm, G4int mode) const {
 
   return reflectedKDir;
 }
+
+
+// Get the position on the edge of two surfaces
+
+G4ThreeVector G4CMPPhononBoundaryProcess::
+GetEdgePosition(const G4ThreeVector& stepLocalPos, const G4ThreeVector& kTan,
+        const G4double stepSize) const {
+  // Get normal at current position
+  G4VSolid* solid = GetCurrentVolume()->GetLogicalVolume()->GetSolid();
+  G4ThreeVector currNorm = solid->SurfaceNormal(stepLocalPos);
+
+  // Step into the normal to get comfortably on the other surface
+  G4ThreeVector edgePos = stepLocalPos - 1.*mm * currNorm;
+
+  // Step back to surface along kTan and correct with currNorm
+  G4double surfAdjust = solid->DistanceToIn(edgePos, -kTan);
+  edgePos -= surfAdjust * kTan;
+  edgePos += 1.*mm * currNorm;
+
+  // Get new wavevector reflected on edge and get kTan
+  G4ThreeVector newKTan = GetReflectionOnEdge(stepLocalPos, kTan);
+
+  // Step remaining distance along surface
+  edgePos += (stepSize - surfAdjust) * newKTan;
+
+  return edgePos;
+}
+
+
+// Reflect "surface mode" phonon at the edge cases
+
+G4ThreeVector G4CMPPhononBoundaryProcess::
```

Comment:
See my comment in the .hh file about passing in non-const reference and modifying in place.